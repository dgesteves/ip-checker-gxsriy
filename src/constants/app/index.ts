export const HEADER_TITLE = "IP Address Tracker";
export const SEARCH_PLACEHOLDER = "Search for any IP address or domain";
export const TIME_STANDARD = "UTC";
export const IPIFY_API_URL = "https://api.ipify.org";
export const WOOKIE_BASE_API_URL = "https://wookie.codesubmit.io/ipcheck";
export const LOCALHOST_IP = "127.0.0.1";
export const MAP_FALLBACK_TEXT = "No location found for this IP address";
export const MAP_FALLBACK_LAT = 0;
export const MAP_FALLBACK_LNG = 0;
export const MAP_ZOOM = 17;
export const VALID_IP_ADDRESS =
  /^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;

export enum INFO_CARD_ELEMENTS {
  IP = "ip address",
  LOCATION = "location",
  TIMEZONE = "timezone",
  ISP = "isp",
}

export enum MAP_TILE_LAYER {
  ATTRIBUTION = '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors',
  URL = "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
}
