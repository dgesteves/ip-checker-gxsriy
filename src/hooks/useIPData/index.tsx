import { useMutation } from "react-query";
import getIPData from "../../services/wookieApi";
import { useEffect, useState } from "react";
import { VALID_IP_ADDRESS } from "../../constants/app";
import { IWookieApiResponse } from "../../services/wookieApi/types";

export default function useIPData(IP: string) {
  const [curData, setCurData] = useState<IWookieApiResponse>();
  const { data, error, isSuccess, isLoading, mutateAsync } =
    useMutation(getIPData);

  useEffect(() => {
    const storageIPData = localStorage.getItem(String(IP));
    if (IP && !storageIPData) {
      mutateAsync(IP).then((data) => {
        if (VALID_IP_ADDRESS.test(IP)) {
          localStorage.setItem(String(IP), JSON.stringify(data));
        }
      });
    } else if (IP && storageIPData) {
      setCurData(JSON.parse(storageIPData));
    }
  }, [IP, mutateAsync]);

  return {
    IPData: data || curData,
    IPIsLoading: isLoading,
    IPError: error,
    IPIsSuccess: isSuccess,
  };
}
