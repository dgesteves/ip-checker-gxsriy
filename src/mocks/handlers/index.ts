import { rest } from "msw";
import { IPIFY_API_URL, WOOKIE_BASE_API_URL } from "../../constants/app";
import {
  DEFAULT_IP_ADDRESS,
  GETIPDATA_RESPONSE,
  GETOWNIP_RESPONSE,
} from "../../constants/tests";

export const handlers = [
  rest.get(IPIFY_API_URL, (req, res, ctx) => {
    req.url.searchParams.set("format", "json");
    return res(ctx.status(200), ctx.json(GETOWNIP_RESPONSE));
  }),
  rest.get(`${WOOKIE_BASE_API_URL}`, (req, res, ctx) => {
    req.url.searchParams.set("ip", DEFAULT_IP_ADDRESS);
    req.headers.set("Authorization", "Bearer WookieIP2022");
    return res(ctx.status(200), ctx.json(GETIPDATA_RESPONSE));
  }),
];
