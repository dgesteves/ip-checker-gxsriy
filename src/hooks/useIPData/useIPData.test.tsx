import { renderHook, waitFor } from "@testing-library/react";
import useIPData from "./index";
import { QueryClient, QueryClientProvider } from "react-query";
import { DEFAULT_IP_ADDRESS, GETIPDATA_RESPONSE } from "../../constants/tests";
import { server } from "../../mocks/server";
import { rest } from "msw";
import { WOOKIE_BASE_API_URL } from "../../constants/app";

describe("useIPData hook", () => {
  const queryClient = new QueryClient();

  const wrapper = ({ children }: { children: JSX.Element | JSX.Element[] }) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );

  it("should return the correct shape", async () => {
    const { result } = renderHook(() => useIPData(DEFAULT_IP_ADDRESS), {
      wrapper,
    });

    await waitFor(() => result.current);

    expect(result.current).toMatchSnapshot();
  });

  it("should return the correct shape when loading", async () => {
    const { result } = renderHook(() => useIPData(DEFAULT_IP_ADDRESS), {
      wrapper,
    });

    await waitFor(() => result.current.IPIsLoading);

    expect(result.current).toMatchSnapshot();
  });

  it("should return the correct shape when error", async () => {
    const { result } = renderHook(() => useIPData(DEFAULT_IP_ADDRESS), {
      wrapper,
    });

    await waitFor(() => result.current.IPError);

    expect(result.current).toMatchSnapshot();
  });

  it("should return the correct shape when success", async () => {
    const { result } = renderHook(() => useIPData(DEFAULT_IP_ADDRESS), {
      wrapper,
    });

    await waitFor(() => result.current.IPIsSuccess);

    expect(result.current).toMatchSnapshot();
  });

  it("should return the correct shape when loading", async () => {
    const { result } = renderHook(() => useIPData(DEFAULT_IP_ADDRESS), {
      wrapper,
    });

    await waitFor(() => result.current.IPIsLoading);

    expect(result.current.IPIsLoading).toEqual(false);
  });

  it("should return the correct shape when error", async () => {
    const { result } = renderHook(() => useIPData(DEFAULT_IP_ADDRESS), {
      wrapper,
    });

    await waitFor(() => result.current.IPError);

    expect(result.current.IPError).toEqual(null);
  });

  it("should return the correct shape when success", async () => {
    const { result } = renderHook(() => useIPData(DEFAULT_IP_ADDRESS), {
      wrapper,
    });

    await waitFor(() => result.current.IPIsSuccess);

    expect(result.current.IPIsSuccess).toEqual(false);
  });

  it("should return the correct values after success api call", async () => {
    server.use(
      rest.get(`${WOOKIE_BASE_API_URL}`, (req, res, ctx) => {
        req.url.searchParams.set("ip", DEFAULT_IP_ADDRESS);
        req.headers.set("Authorization", "Bearer WookieIP2022");
        return res(ctx.status(200), ctx.json(GETIPDATA_RESPONSE));
      })
    );

    const { result } = renderHook(() => useIPData(DEFAULT_IP_ADDRESS), {
      wrapper,
    });

    await waitFor(() => {
      expect(result.current.IPData).toEqual(GETIPDATA_RESPONSE);
      expect(result.current.IPIsLoading).toBe(false);
      expect(result.current.IPError).toBe(null);
      expect(result.current.IPIsSuccess).toBe(true);
    });
  });
});
