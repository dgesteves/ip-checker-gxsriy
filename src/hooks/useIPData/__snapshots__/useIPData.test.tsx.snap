// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useIPData hook should return the correct shape 1`] = `
Object {
  "IPData": undefined,
  "IPError": null,
  "IPIsLoading": false,
  "IPIsSuccess": false,
}
`;

exports[`useIPData hook should return the correct shape when error 1`] = `
Object {
  "IPData": undefined,
  "IPError": null,
  "IPIsLoading": false,
  "IPIsSuccess": false,
}
`;

exports[`useIPData hook should return the correct shape when loading 1`] = `
Object {
  "IPData": undefined,
  "IPError": null,
  "IPIsLoading": false,
  "IPIsSuccess": false,
}
`;

exports[`useIPData hook should return the correct shape when success 1`] = `
Object {
  "IPData": undefined,
  "IPError": null,
  "IPIsLoading": false,
  "IPIsSuccess": false,
}
`;
